### PR TITLE
Migrate to Strapi v5

### PR DESCRIPTION
**Pull request by mattmilburn**
## 🚧 Work In Progress 🚧

This PR brings support for [Strapi v5](https://strapi.io/five) to the plugin.

* Use 18-20 range for `nodeEngines`.
* Upgrade dependencies to their latest versions.
* Replace `plugin.name` with `plugin::name` in server config service.
* Refactor plugin structure for v5.
* Refactor to use TypeScript.

### CODE REVIEW

**Comment by SalahAdDin** [comment]
Waiting for this! Thank you for this great plugin.

**Comment by SalahAdDin** [comment]
Does a "simple" plugin require that much?

**Comment by jackbuehner** [comment]
@SalahAdDin Strapi made a lot of changes for version 5 that are still undocumented. It takes quite a bit of sleuthing to find the new way to do things, and it has only recently become worth it now that we are in later release candidates.

**Comment by SalahAdDin** [comment]
> @SalahAdDin Strapi made a lot of changes for version 5 that are still undocumented. It takes quite a bit of sleuthing to find the new way to do things, and it has only recently become worth it now that we are in later release candidates.

Oh man, don't tell me that, I am planning to help to migrate a plugin.

**Comment by danidoff** [comment]
Hey @mattmilburn, any news around this ? Do you know when it's going to be merged this and released ?

**Comment by SalahAdDin** [comment]
@danidoff it is taking time for everyone to migrate from one to another version.

File: admin/src/components/EditViewRightLinks/index.tsx [hunk 1]
```diff
@@ -1,4 +1,3 @@
-import React from 'react';
 import { useCMEditViewDataManager } from '@strapi/helper-plugin';
```
Comment (Sagar-echo) [inline admin/src/components/EditViewRightLinks/index.tsx:1]: strapi/helper-plugin is deprecated - https://docs.strapi.io/dev-docs/migration/v4-to-v5/additional-resources/helper-plugin

Comment (mattmilburn) [reply admin/src/components/EditViewRightLinks/index.tsx:1]: @Sagar-echo There is currently an [open PR to fix this](https://github.com/mattmilburn/strapi-plugin-preview-button/pull/140), but so far it has not been resolved

Comment (SalahAdDin) [reply admin/src/components/EditViewRightLinks/index.tsx:1]: > strapi/helper-plugin is deprecated - https://docs.strapi.io/dev-docs/migration/v4-to-v5/additional-resources/helper-plugin

is it deprecated? really? why?

**Comment by michsko544** [comment]
> @danidoff Currently no specific date for release, but we are working on it. And despite Strapi v5 being officially released, it is still a work in progress with bug fixes, updates, etc. still coming through.
> 
> Happy to get some help if you can provide any. We're currently stuck wondering why we can't use Strapi hooks in the `EditViewRightLinks` component without getting a `___ must be used within StrapiApp` error.

Maybe `___ must be used within StrapiApp` it is related to inconsistent versions of Strapi packages like here: https://github.com/strapi/strapi/issues/21779
I had a similar error and solved it by forcing to install the same versions of packages using `overrides` directive

**Comment by mattmilburn** [comment]
This is still a problem in v5.4.0. I've created [an issue with Strapi](https://github.com/strapi/strapi/issues/22162) to describe the problem in case anyone is interested.

**Comment by mattmilburn** [comment]
Not sure what happened but this was not meant to merge to staging when I merged another brach into this one. Going to remake this PR and finalize this update.

**Pull request merged**